packages/rich-text-structure/src/create.js

```diff
@@ -0,0 +1,326 @@
+/**
+ * Internal dependencies
+ */
+
+import { isEmpty } from './is-empty';
+
+/**
+ * Browser dependencies
+ */
+
+const { TEXT_NODE, ELEMENT_NODE } = window.Node;
+
+/**
+ * Parse the given HTML into a body element.
+ *
+ * @param {string} html The HTML to parse.
+ *
+ * @return {HTMLBodyElement} Body element with parsed HTML.
+ */
+function createElement( html ) {
+	const htmlDocument = document.implementation.createHTMLDocument( '' );
+
+	htmlDocument.body.innerHTML = html;
+
+	return htmlDocument.body;
+}
+
+/**
+ * Creates rich text value and selection objects from a DOM element and range.
+ *
+ * @param {HTMLElement} element      Element to create value object from.
+ * @param {Range}       range        Range to create selection object from.
+ * @param {string}      multilineTag Multiline tag if the structure is multiline.
+ * @param {Object}      settings     Settings passed to `createRecord`.
+ *
+ * @return {Object} A rich text record.
+ */
+export function create( element, range, multilineTag, settings ) {
+	if ( typeof element === 'string' ) {
+		element = createElement( element );
+	}
+
+	if ( ! multilineTag ) {
+		return createRecord( element, range, settings );
+	}
+
+	const emptyRecord = {
+		formats: [],
+		text: '',
+	};
+
+	if ( ! element || ! element.hasChildNodes() ) {
+		return emptyRecord;
+	}
+
+	const children = Array.from( element.children )
+		.filter( ( { nodeName } ) => nodeName.toLowerCase() === multilineTag );
+	const maxIndex = children.length - 1;
+
+	if ( maxIndex < 0 ) {
+		return emptyRecord;
+	}
+
+	return children.reduce( ( accumulator, node, index ) => {
+		const { start, end, text, formats } = createRecord( node, range, settings );
+		const length = accumulator.text.length;
+
+		if ( range ) {
+			if ( start !== undefined ) {
+				accumulator.start = length + start;
+			} else if (
+				node.parentNode === range.startContainer &&
+				node === range.startContainer.childNodes[ range.startOffset ]
+			) {
+				accumulator.start = length;
+			}
+
+			if ( end !== undefined ) {
+				accumulator.end = length + end;
+			} else if (
+				node.parentNode === range.endContainer &&
+				node === range.endContainer.childNodes[ range.endOffset - 1 ]
+			) {
+				accumulator.end = length + text.length;
+			}
+		}
+
+		accumulator.formats = accumulator.formats.concat( formats );
+		accumulator.text += text;
+
+		if ( index !== maxIndex ) {
+			accumulator.formats = accumulator.formats.concat( [ , , ] );
+			accumulator.text += '\n\n';
+		}
+
+		return accumulator;
+	}, emptyRecord );
```

Comment:
fun fact!

at the end of this call `emptyRecord` is no longer `emptyRecord` - `emptyRecord` is the final value of the reduce.

normally we wouldn't modify the accumulator while iterating. instead we'd return a new accumulator for each iteration. it's JavaScript so 🤷‍♂️ whatever, but it may surprise some people seeing the mutation like this.

for clarity we could return a new one, for performance we could possibly skip the reduce. if we want testability then the reduce is still lovely but we could additionally extract the inner function

```js
const extendStart = ( { length, start }, { start: nextStart }, node, range ) => {
	// when given a new range add by its length or something
	if ( nextStart !== undefined ) {
		return length + nextStart;
	}

	// we might be a child of the parent?
	if (
		node.parentNode === node.startContainer &&
		node === range.startContainer.childNodes[ range.startOffset ]
	) {
		return length;
	}

	// otherwise stay put
	return start;
}

const extendEnd = ( { length }, { end, text }, node, range ) => {
	if ( end !== undefined ) {
		return length + end;
	}

	if (
		node.parentNode === range.endContainer &&
		node === range.endContainer.childNodes[ range.endOffset - 1 ]
	) {
		return length + text.length;
	}
}

const extend = ( prev, node, range, [ suffixText, suffixFormats ] = [ '', [] ] ) => {
	const record = createRecord( node, range, settings );

	return {
		start: range ? extendStart( prev, record, node, range ) : prev.start,
		end: range ? extendEnd( prev, record, node, range ) : prev.end,
		text: prev.text + record.text + suffixText,
		formats: [ ...prev.formats, ...record.formats, ...suffixFormats ],
	};
};

return children.reduce( ( record, child, index ) =>
	extend( record, child, range, index !== maxIndex ? [ '\n\n', [ , , ] ] ),
	emptyRecord
);
```

by rewriting init this way it allows us to more granularly test the behavior, which I suppose is pretty critical while building the structure.

```js
test( 'extendStart', () => {
	it( 'should move forward when we have a new start position', () => {
		const prev = createRecord( some, values, whatever );
		const next = createRecord( some, other, values );

		expect( extendStart( prev, next, null, null ) ).toBe( prev.length + next );
	} )

	it( 'should do something else if some node relationship exists', () => {
		const prev = createRecord( … )
		const next = createSomeRelationshipRecord( … );

		expect( extendStart( prev, next, node, range ) ).toBe( prev.length );
	} )

	if ( 'should have no change for no range', () => {
		expect( { start: 4 }, { start: undefined }, node, null ).toBe( 4 )
	} )
} )

test( 'extendEnd', () => … );

test( 'extend', () => {
	it( 'should join the text', () => {
		const emptyJoin = extend( emptyRecord, plainNode, null, null );
		const nextJoin = extend( emptyJoin, plainNode, null, null );

		expect( emptyJoin ).toHaveProperty( 'text', plainNode.text );
		expect( nextJoin ).toHaveProperty( 'text', plainNode.text + plainNode.text );
	} )

	…
} );
```

blah blah blah - those are bad tests but I hope it demonstrates breaking it down into testable units

Reply:
Does it matter if it's no longer used after that? In other words, does it matter if it's added there for that reason, to be modified? What's problematic about it?

Reply:
the main problem is that it is likely to surprise devs who come in not expecting it. one can say that it's a purity argument but the point of the argument is that we're somewhat adopting a well-defined standard (the reducer) but using it in a way that breaks the grounding assumptions.

typically I wouldn't throw much object if we're limited to the reduction, but in this case have a subtle behavior that's not obvious: we're mutating `emptyRecord` which is defined many lines above.

practical suggested resolution?
 - define `emptyRecord` inline as `reduce( f, { text: '', formats: [] } )`
 - or clone it to make it clear we're likely mutating it `reduce( f, { ...emptyRecord } )`